Added log parsing to support colorized build logs

DreierF opened the pull request:
Hey,

I added ANSI code parsing to the build details in order to make the logs more readable (#7).
travis_... commands are removed at least visually.
![build-log](https://user-images.githubusercontent.com/5631865/29833767-c8e5b192-8cec-11e7-8357-e417fd04989b.png)

dkhmelenko commented:
@DreierF I tested solution, it looks nice! 👍 
However, I would like to improve a bit codebase. Wouldn't you mind if in the next days I will use the code from this PR and prepare another PR (because I cannot commit right to your fork)?
I will also provide my feedback here, what I want to improve.

DreierF commented:
Sure. 
Feedback is always appreciated 🙂

dkhmelenko commented on app/src/main/java/com/khmelenko/lab/varis/log/TextLeaf.java:
```diff
@@ -0,0 +1,148 @@
+        }
+    }
+
+    private void translateAnsiCode(String code) {
```
This function looks more like utililities function, I would move it to once of the utilities classes

DreierF replied:
You are right, I would move it to the AnsiParser class if we don't move the colors to colors.xml

dkhmelenko commented on app/src/main/java/com/khmelenko/lab/varis/log/LogEntryComponent.java:
```diff
@@ -0,0 +1,7 @@
+package com.khmelenko.lab.varis.log;
+
+public abstract class LogEntryComponent {
```
what was the reason to have this as an abstract class and not interface?

DreierF replied:
No special reason behind it. I can change it to an interface as well.

dkhmelenko commented on app/src/main/java/com/khmelenko/lab/varis/log/TravisCommandLeaf.java:
```diff
@@ -0,0 +1,9 @@
+package com.khmelenko.lab.varis.log;
+
+public class TravisCommandLeaf extends LogEntryComponent {
+    public String command;
```
I would prefer to access this field via getter

DreierF replied:
Sure

dkhmelenko commented:
Let me know what do you think about my comments

DreierF commented on app/src/main/java/com/khmelenko/lab/varis/log/AnsiColor.java:
```diff
@@ -0,0 +1,29 @@
+package com.khmelenko.lab.varis.log;
+
+public class AnsiColor {
```
I can do that, but then you need to load every value with getColor(), which means you have to pass a Context and hence make the parser dependant on the Context.

DreierF commented on app/src/main/java/com/khmelenko/lab/varis/log/AnsiParser.java:
```diff
@@ -0,0 +1,96 @@
+ * */
+public class AnsiParser {
+
+    private Stack<TextLeaf> result = new Stack<>();
```
You are right, that's no longer necessary. It was used before to make the [0K parsing work. For the case when there was no line break in the current TextLeaf I had to go back in the Stack and delete every until I get a line break. But [0K seems to only appear directly after a line break in Travis, which rendered it useless so I removed it. What would you suggest, fully implement it and continue using stack or get rid of the stack?

DreierF commented on app/src/main/java/com/khmelenko/lab/varis/log/AnsiParser.java:
```diff
@@ -0,0 +1,96 @@
+
+    private Stack<TextLeaf> result = new Stack<>();
+    private TextLeaf textLeaf = new TextLeaf();
+    private String[] ansiStates = new String[]{};
```
Ups, sorry. Will change that.

DreierF commented on app/src/main/java/com/khmelenko/lab/varis/log/AnsiParser.java:
```diff
@@ -0,0 +1,96 @@
+    private Stack<TextLeaf> parse(String text) {
+        // Remove character when followed by a BACKSPACE character
+        while (text.contains("\b")) {
+            text = text.replaceAll("^\b+|[^\b]\b", "");
```
Yes, that's a bug in Android studio concerning \b character in Regexes, but I checked it. It's valid.

DreierF replied:
https://stackoverflow.com/questions/42537321/intellij-says-b-backspace-is-an-illegal-escape-sequence-inside-a-string-liter

DreierF commented:
I implemented all of your suggestions except moving the colors to `colors.xml` (see my comment above). I also did extract the formatting attributes into a separate class `FormattingOptions` and optimized the `TravisCommandLeaf` class.
Let me know what you think 😉
P.S. Not sure if you have seen it, but I've also updated the other pull request regarding the started animation state.

dkhmelenko commented on app/src/main/java/com/khmelenko/lab/varis/log/FormattingOptions.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    public void setTextColor(Integer textColor) {
+        this.mTextColor = textColor;
```
don't need to have this, sice we have `m` prefix for fields :)

DreierF replied:
Of course 😄

dkhmelenko commented on app/src/main/java/com/khmelenko/lab/varis/log/LogEntryComposite.java:
```diff
@@ -0,0 +1,33 @@
+public class LogEntryComposite implements LogEntryComponent {
+
+    private String name;
+    private List<LogEntryComponent> logEntryComponents = new ArrayList<>();
```
`m` prefix is missing for fields

dkhmelenko commented on app/src/main/java/com/khmelenko/lab/varis/log/TravisCommands.java:
```diff
@@ -0,0 +1,83 @@
+        return Observable.fromIterable(logEntryComponents)
+                .flatMap(logEntryComponent -> Observable.fromIterable(splitAtTravisCommands(logEntryComponent)))
+                .toList()
+                .blockingGet();
```
`blockingSingle()` should be better in this case, because there is only 1 item in Single and it's a List

DreierF replied:
Not sure if I get what you mean. It's on purpose to get a List here...

dkhmelenko replied:
ah, yeah, sorry, my mistake, you are right here

dkhmelenko commented on app/src/main/java/com/khmelenko/lab/varis/log/TravisCommands.java:
```diff
@@ -0,0 +1,83 @@
+            if (logEntryComponent instanceof TravisCommandLeaf) {
+                TravisCommandLeaf command = (TravisCommandLeaf) logEntryComponent;
+                if (command.getCommand().equals(TRAVIS_FOLD)) {
+                    if (command.getType().equals(START)) {
```
This is alread a 4th level of nested operators. Try to limit it by having max 2 levels of nexted operators by splitting in functions.

dkhmelenko commented:
P.S.: I've seen another PR for fixing animation. I prefer to complete with this one and then move forward to that one :)

DreierF commented:
I fixed all the stuff you commented on
